### PR TITLE
Enable Mapping Save button when Valid

### LINKED ISSUE #2808: Mapping Page: Save button should enable as soon as any valid text is in the field, not after leaving the field
Right now, you have to click outside the text box after entering an org before the save button is enabled. The button should be enabled as soon as valid data is present in the box.

delayed_save

### PR DESCRIPTION
**Pull request by ischolten**
- [x] CHANGELOG.md updated with a link to the PR (not the Issue)
  - [x] Rebased/mergable
  - [x] Tests pass

Resolves #2808 

### The problem
InputClickToEdit was only calling the update function during onBlur.

### The Solution
Create another update prop function for onKeyUpdate.

### CODE REVIEW

File: ui/src/shared/components/InputClickToEdit.js [hunk 1]
```diff
@@ -88,7 +96,8 @@ const {func, bool, number, string} = PropTypes
 InputClickToEdit.propTypes = {
   wrapperClass: string.isRequired,
```
Comment (ebb-tide) [inline ui/src/shared/components/InputClickToEdit.js]: You could pull `<div className={wrapperClass}>` outside the if statement in the render function to stay DRY. :)

File: ui/src/shared/components/InputClickToEdit.js [hunk 2]
```diff
@@ -63,7 +71,7 @@ class InputClickToEdit extends Component {
                 className="form-control input-sm provider--input"
                 defaultValue={value}
                 onBlur={this.handleInputBlur}
-                onKeyDown={this.handleKeyDown}
+                onKeyUp={this.handleKeyUp}
                 autoFocus={true}
                 onFocus={this.handleFocus}
                 ref={r => (this.inputRef = r)}
```
Comment (ebb-tide) [inline ui/src/shared/components/InputClickToEdit.js]: It doesn't look like you're using inputRef anymore?

Comment (ebb-tide) [reply ui/src/shared/components/InputClickToEdit.js]: if tabIndex is not specified in props, then does the tabIndex of the input become undefined? could it default to 0? so the input would behave as expected? https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex

File: ui/src/shared/components/InputClickToEdit.tsx [hunk 3]
```diff
@@ -1,34 +1,67 @@
...
     }
 
-    this.setState({isEditing: false, value: e.target.value})
+    this.setState({
+      isEditing: false,
+      initialValue: e.target.value,
+    })
   }
 
   handleKeyDown = e => {
```
Comment (121watts) [inline ui/src/shared/components/InputClickToEdit.tsx]: Lets define these event types also

File: ui/src/shared/components/InputClickToEdit.tsx [hunk 4]
```diff
@@ -1,34 +1,67 @@
-import React, {Component, PropTypes} from 'react'
+import React, {PureComponent} from 'react'
 
-class InputClickToEdit extends Component {
+interface Props {
+  wrapperClass: string
+  value?: string
+  onChange?: (value: string) => void
+  onBlur: (value: string) => void
```
Comment (121watts) [inline ui/src/shared/components/InputClickToEdit.tsx]: This `onChange` and `onBlur` probably take `Event` types.

File: ui/test/shared/components/InputClickToEdit.test.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        const isEditing = inputClickToEdit.state('isEditing')
+
+        expect(isEditing).toBe(true)
+        expect(divAfterClick.exists()).toBe(false)
+        expect(inputField.exists()).toBe(true)
+      })
+    })
+  })
+
+  describe('instance methods', () => {})
```
Comment (121watts) [inline ui/test/shared/components/InputClickToEdit.test.tsx]: You can cut this out.  😄

File: ui/test/shared/components/ProvidersTableRowNew.test.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  const row = shallow(<ProvidersTableRowNew {...props} />)
+  return {
+    props,
+    defaultState,
+    row,
+  }
+}
+
+describe('Components.Shared.ProvidersTableRowNew', () => {
+  describe('rendering', () => {})
```
Comment (121watts) [inline ui/test/shared/components/ProvidersTableRowNew.test.tsx]: Let's cut anything that isn't used.

File: ui/test/shared/components/ProvidersTableRowNew.test.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+        const confirmButtons = row.find(ConfirmButtons)
+        const confirmButtonsDisabled = confirmButtons.prop('isDisabled')
+
+        expect(confirmButtonsDisabled).toBe(false)
+      })
+    })
+  })
+
+  describe('instance methods', () => {})
```
Comment (121watts) [inline ui/test/shared/components/ProvidersTableRowNew.test.tsx]: same here :)

File: ui/src/shared/components/InputClickToEdit.tsx (hunk 4 above)
Comment (ischolten) [reply ui/src/shared/components/InputClickToEdit.tsx]: They do take strings types. I can change the name of those props if they are too similar to the props on the HTML input

**Pull request merged**